### ISSUE TITLE
Retrieve type of unmanaged HeapAlloc events

### ISSUE THREAD

**Issue by kayle** [microsoft/perfview#780]
PerfView currently supports tracking native allocations/stacks.

I investigated what it would take to output the types of allocations as well (e.g. int[], MyStruct, etc). I believe Visual Studio is using the same ETW events (https://docs.microsoft.com/en-us/visualstudio/profiling/memory-usage?view=vs-2017), but it is non-obvious to me how to get from AllocAddress + pdb to a c++ type.

**Comment by vancem** [comment]
Indeed it does sound like VS knows something that PerfView does not with respect to native allocation instrumentation (since it can get the type of the allocation, and PerfView can't).   

I would start by making a very simple demo (that say allocate a lot of just a couple paraticular types in native code).   Just having a PerfView CPU trace (use /SessionName=MySession and a PerfView Ver > 2.0.27 to avoid conflicts over the kernel session), will be useful to determine what they are doing.    If we can look at their ETL file even better.  

Most likely there is either 
1. more information being logged at allocation time. OR
2. It is deducing it from the allocation stack (possibly by having addition PDB data).

**Comment by kayle** [comment]
There is an etl file created by Visual Studio, and it shows HeapAlloc events +stacks as the primary info logged, so I believe (2) the pdb is supplying the missing type information and not (1). 

This header file mentioned [here](https://blogs.msdn.microsoft.com/vcblog/2016/05/25/tracking-custom-memory-allocations-with-visual-studio-15-preview-2/) allows collecting the same info for a custom heap:
https://msdnshared.blob.core.windows.net/media/2016/05/VSCustomNativeHeapEtwProvider.h

Is there someone familiar with the c++ pdb/visual studio interaction that could chime in?

**Comment by kayle** [comment]
I should've read the previously linked page more closely. It specifies exactly how it works:

> At each call to an allocator function, the address of the call site, the size of the call instruction, and the typeid of the new object are emitted into a new S_HEAPALLOCSITE symbol inside the PDB.  When the Windows heap code emits an ETW event with a call stack for an allocation, the memory tool walks the call stack looking for a return address matching an S_HEAPALLOCSITE symbol.  The typeid in the symbol determines the runtime type of the allocation.

Using the data referenced from https://github.com/KirillOsenkov/Dia2Dump/blob/master/PrintSymbol.cpp
```
////////////////////////////////////////////////////////////
// Print a callsite symbol info: SEG:OFF, RVA, type
//
void PrintHeapAllocSite(IDiaSymbol *pSymbol)
{
  DWORD dwISect, dwOffset;
  if (pSymbol->get_addressSection(&dwISect) == S_OK &&
    pSymbol->get_addressOffset(&dwOffset) == S_OK) {
    wprintf(L"[0x%04x:0x%08x]  ", dwISect, dwOffset);
  }

  DWORD rva;
  if (pSymbol->get_relativeVirtualAddress(&rva) == S_OK) {
    wprintf(L"0x%08X  ", rva);
  }

  IDiaSymbol *pAllocType;
  if (pSymbol->get_type(&pAllocType) == S_OK) {
    PrintType(pAllocType);
    pAllocType->Release();
  }
}
```

The NativeSymbolModule's IDiaSession can be used to read this information:

```
            m_session.getHeapAllocationSites(out var diaEnumSymbols);
            for (; ; )
            {
                diaEnumSymbols.Next(1, out var sym, out var fetchCount);
                if (fetchCount == 0)
                {
                    break;
                }

                var section = sym.addressSection;
                var offset = sym.addressOffset;
                var rva = sym.relativeVirtualAddress;

                m_session.symbolById(sym.typeId, out var typeSym);
                var typeName = typeSym?.name;

```

Would you be interested in a PR to add this functionality? Would it go in the current "Net OS Heap Alloc Stacks" or be it's own view? If I were picking, I think I'd either want two native memory views that correspond to the managed GC Heap Net Mem/GC Heap Alloc Ignore Free, or alternatively just rename the GC Heap views to not be managed specific and have the native allocations show up there if the information was available.

**Comment by brianrob** [comment]
Fixed via #857.
